Mathlib/Topology/Algebra/FilterBasisNew.lean

```diff
@@ -0,0 +1,448 @@
+/-
+Copyright (c) 2021 Patrick Massot. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Patrick Massot, Anatole Dedecker
+-/
+import Mathlib.Order.Filter.Bases
+import Mathlib.Topology.Algebra.Module.Basic
+
+/-!
+**WARNING** : This is a new version of `Topology.Algebra.FilterBasis`, which is still used
+by the library at this point in time. The library will be gradually modified to use this file
+instead. See `https://leanprover.zulipchat.com/#narrow/channel/287929-mathlib4/topic/Refactoring.20algebraic.20filter.20bases/near/479437345`
+for more details.
+
+# Algebraic filter bases
+
+We know that the topology of a topological group (or ring, or module) is invariant by
+translations, hence it is completely described by its filter of neighborhoods at the
+neutral element (`1` for groups, `0` for additive groups, rings and modules).
+In this file, we completely characterize the filter bases which can be obtained as bases of
+neighborhoods of the neutral element for some group (or ring, or module) topology.
+
+More precisely, for `X` in `{Group, AddGroup, Ring, Module}`, we provide :
+* `Filter.IsXBasis p s`: a strenghtening of `Filter.IsBasis` with compatibility conditions with the
+  algebraic structure.
+* `Filter.IsXBasis.mk_of_comm`: simpler contructors in the commutative setting.
+* `Filter.HasBasis.isXBasis`: in a topological `X`, *any* basis of neighborhoods
+  of the neutral element satisfies `Filter.IsXBasis`.
+* `Filter.IsXBasis.topologicalX_of_hasBasis`: if some `X` comes *equipped* with a topology and
+  a basis of neighborhoods of the neutral element satisfies `Filter.IsXBasis`, then that topology is
+  a `X` topology. This requires the topology to be invariant by translation.
+* `Filter.IsXBasis.topology`: the topological structure associated to some `X` filter basis.
+* `Filter.IsXBasis.topologicalX`: the topology associated to some `X` filter basis is indeed
+  an `X` topology.
+
+Note the subtlety between `IsXBasis.topologicalX_of_hasBasis`, useful for *proving* that some
+topology is compatible with the algebraic structure, and `IsXBasis.topologicalX` which shows
+that *the* topology associated to some basis is compatible with the algebraic structure.
+Mathematically this is the same thing, but having these variations is convenient.
+
+Remark: for `X = Module`, we use `continuousSMul` instead of `topologicalModule` in the names.
+
+## Usage indications
```

Comment:
The docstring is quite detailed but it might be enhanced by two or three baby case examples that show how to use it in practice. Possibly a link to the relevant mathlib files is enough, in which case this could be done afterwards, once this file has became the canonical API.

Reply:
Yes, I think that linking a posteriori to other files is the best approach. I may rewrite the "usage indications" a bit though.